tools/autograd/gen_variable_type.py

```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+    # de-duplicate operation names
+    functions_by_name = defaultdict(list)
+    for func in autograd_functions:
+        functions_by_name[func['op']].append(func)
+    for op in sorted(functions_by_name.keys()):
+        overloads = functions_by_name[op]
+        if len(overloads) > 1:
+            for i, func in enumerate(overloads):
+                func['op'] += str(i)
+
+
+def preprocess_nn_functions(declarations):
+    declarations_by_name = defaultdict(list)
+    for d in declarations:
+        declarations_by_name[d['name']].append(d)
+
+    autograd_functions = []
+    for declaration in declarations:
+        name = declaration['name']
+        if name == 'batch_norm' or 'conv' in name:
+            continue
+
+        fwd_name = name + '_forward'
+        if fwd_name not in declarations_by_name:
+            continue
+        declaration['base_name'] = fwd_name
 
-        saved = []
-        for arg in option['python_arguments']:
+        fwd = declarations_by_name[name + '_forward'][0]
+
+        input_num = 0
+        bwd_name = name + '_backward'
+        assert len(declarations_by_name[bwd_name]) == 1
+        bwd = declarations_by_name[bwd_name][0]
+
+        def actual(arg):
             name = arg['name']
-            sizes_name = name + '_sizes'
-            if any(re.search(name_regex.format(name), f) for f in derivatives):
-                saved.append(arg)
-            if any(sizes_name in f for f in derivatives):
-                saved.append({
-                    'name': sizes_name,
-                    'type': 'IntList',
-                })
-            for f in derivatives:
-                for match_name in re.findall(r"{}_argsize_\w+".format(name), f):
-                    saved.append({
-                        'name': match_name,
-                        'type': 'int64_t',
-                    })
-                for match_name in re.findall(r"{}_argsizes_\w+".format(name), f):
-                    saved.append({
-                        'name': match_name,
-                        'type': 'IntList',
-                    })
-        option['saved'] = saved
-
-        options.append(option)
-
-    options = sorted(options, key=lambda o: o['name'])
-    for name, overloads in groupby(options, lambda o: o['name']):
-        overloads = list(overloads)
-        for i, option in enumerate(overloads):
-            name = option['name']
-            option['op'] = name[0].upper() + name[1:] + 'Backward'
-            if len(overloads) > 1:
-                option['op'] += str(i)
-
-    return options
-
-
-def create_autograd_functions(top_env, declarations):
+            return name if name != 'inplace' else 'false'
+
+        actuals = [actual(arg) for arg in bwd['arguments']]
+        formula = '{}({})'.format(bwd_name, ', '.join(actuals))
+        formula = formula.replace('grad_output', 'grad')
+        if not re.search(IDENT_REGEX.format('grad'), formula):
+            formula = '({}).mul_(grad)'.format(formula)
```

Comment:
Uhhh this makes me nervous. Maybe let's just raise an error if `grad` doesn't appear anywhere

Reply:
Yes, but why can't we include that in the formula itself? Plus `mul_` with 2 tensors is going to raise an error in backward

Reply:
I'm not sure what you mean. These four lines are constructing the formula (from the signature of the backwards function) for the first derivative of NN functions.  The formulas aren't specified anywhere else: it's implicit from the naming of xxx_forwards and xxx_backwards 

(The second derivatives are handled in derivatives.yaml).

Reply:
Ohhhhh I see. Makes sense now. Thank you